clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -66,145 +213,189 @@ class device_vector {
   }
+
   device_vector(device_vector &&other)
-      : _alloc(get_default_queue()), _size(other.size()),
+      : _alloc(std::move(other._alloc)), _size(other.size()),
         _capacity(other.capacity()), _storage(other._storage) {
     other._size = 0;
-    other._capacity = 0; 
+    other._capacity = 0;
     other._storage = nullptr;
   }
 
-  template <typename InputIterator>
-  device_vector(InputIterator first,
-                typename ::std::enable_if<
-                    internal::is_iterator<InputIterator>::value &&
-                        !::std::is_pointer<InputIterator>::value &&
-                        ::std::is_same<typename ::std::iterator_traits<
-                                         InputIterator>::iterator_category,
-                                     ::std::random_access_iterator_tag>::value,
-                    InputIterator>::type last)
-      : _alloc(get_default_queue()) {
-    _size = ::std::distance(first, last);
-    _set_capacity_and_alloc();
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                first, last, begin());
-    }
+  device_vector(device_vector &&other, const Allocator &alloc)
+      : _alloc(alloc), _size(other.size()), _capacity(other.capacity()) {
+    _storage = alloc_traits::allocate(_alloc, _capacity);
+    _construct(other.begin(), other.end()); // ok to parallelize
+    other._size = 0;
+    other._capacity = 0;
+    other._storage = nullptr;
   }
 
   template <typename InputIterator>
-  device_vector(InputIterator first,
-                typename ::std::enable_if<::std::is_pointer<InputIterator>::value,
-                                        InputIterator>::type last)
-      : _alloc(get_default_queue()) {
+  device_vector(
+      InputIterator first,
+      typename ::std::enable_if_t<
+          dpct::internal::is_iterator<InputIterator>::value, InputIterator>
+          last,
+      const Allocator &alloc = Allocator(get_default_queue()))
+      : _alloc(alloc) {
     _size = ::std::distance(first, last);
     _set_capacity_and_alloc();
-    if (_size > 0) {
-      auto ptr_type = sycl::get_pointer_type(first, get_default_context());
-      if (ptr_type != sycl::usm::alloc::host &&
-          ptr_type != sycl::usm::alloc::unknown) {
-        ::std::copy(
-            oneapi::dpl::execution::make_device_policy(get_default_queue()),
-            first, last, begin());
-      } else {
-        sycl::buffer<typename ::std::iterator_traits<InputIterator>::value_type,
-                     1>
-            buf(first, last);
-        auto buf_first = oneapi::dpl::begin(buf);
-        auto buf_last = oneapi::dpl::end(buf);
-        ::std::copy(
-            oneapi::dpl::execution::make_device_policy(get_default_queue()),
-            buf_first, buf_last, begin());
-      }
-    }
+    // unsafe to parallelize on device as we dont know if InputIterator is valid
+    // oneDPL input type
+    _construct_host(first, last);
```

Comment:
If user defined allocator is used and the allocated memory can only be accessed by device, it will not workable. Should we consider such cases?

Reply:
We don't support those cases.  The underlying premise which enables `device_vector` to work is that we are relying on shared USM memory.  If we move away from that, it no longer works generally.

The USM_none case relies on `sycl::buffers`, (and we don't support the same functionality in that case).